rules/Removing/NodeManipulator/ComplexNodeRemover.php

```diff
@@ -57,46 +62,122 @@ public function removePropertyAndUsages(Property $property, array $classMethodNa
                 return;
             }
 
-            $propertyName = (string) $this->nodeNameResolver->getName($propertyFetch);
-            $propertyNames[] = $propertyName;
-            $assigns[$propertyName][] = $assign;
+            $assigns[] = $assign;
         }
 
-        $this->processRemovePropertyAssigns($assigns, $propertyNames);
+        $this->processRemovePropertyAssigns($assigns);
 
         if ($shouldKeepProperty) {
             return;
         }
 
-        // remove __construct param
-
-        /** @var Property $property */
         $this->nodeRemover->removeNode($property);
+    }
 
-        foreach ($property->props as $prop) {
-            if (! $this->nodesToRemoveCollector->isNodeRemoved($prop)) {
-                // if the property has at least one node left -> return
-                return;
+    /**
+     * @param Assign[] $assigns
+     */
+    private function processRemovePropertyAssigns(array $assigns): void
+    {
+        foreach ($assigns as $assign) {
+            // remove assigns
+            $this->assignRemover->removeAssignNode($assign);
+            $this->removeConstructorDependency($assign);
+        }
+    }
+
+    private function isPropertyNameInNewCurrentClassNameSelfClone(string $propertyName, ?ClassLike $classLike): bool
```

Comment:
Could you extract this to own service? The logic seems quite complex

Reply:
implemented, I moved to new service `ForbiddenPropertyRemovalAnalyzer` https://github.com/rectorphp/rector-src/pull/1215/commits/12bf6944360c1ca17b855f6a56290f11a8429b87